### DISCUSSION TITLE
A simple example program to query an iModel

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/itwinjs-core#3483] [Q&A] [answered]
Hello 
Please could you provide a simple example script to open an iModel on iModelHub and perform an ecsql query on the iModel.

I am looking for a simple example node.js script to remotely open an imodel using its ContextId and iModelId and then run an ecsql query on the iModel and output the query result.

**Comment by ramanujam-raman-zz** [comment]
Are you looking to do this in a context of a Web (SPA) at the frontend, or at the backend (of a Web/Desktop application)? Also, do you have authorization figured out, or do you need a sample that included that as well?

**Comment by PeterJBassett** [comment]
Thank you very much for your reply. I am looking for a minimal node.js script with no user interface. It would use the @itwin/oidc-signin-tool to do the sign-in authorisation. It's just that I can't see how to open an iModel database using its ContextId and iModelId, I have spent a long time searching the documentation, but can't see how to do that, and then do a simple ecsql query of the iModel.

**Comment by mgooding** [comment]
I had a version of the program from this [blog post](https://medium.com/itwinjs/writing-command-line-applications-with-imodel-js-ba06f4531cc2) that did exactly this, but it no longer functions due to numerous API changes in authentication and briefcase management.

When [this PR](https://github.com/iTwin/auth-clients/pull/69) goes through, we'll have a command-line authentication solution again and I'll be able to provide the example you're looking for. This is very similar to what [imodel-unity-example](https://github.com/iTwin/imodel-unity-example/blob/master/node/src/IModelHubDownload.ts) does (note that the linked code is for iTwin.js 2.x until the referenced PR goes through).

**Reply by mgooding** [reply]
OK, here's the example as promised: https://github.com/mgooding/simple-itwinjs-cli-app

Try the instructions in the README and let me know how it goes.

**Comment by PeterJBassett** [comment, marked as answer]
Thank you so much for the example script. It worked perfectly for me.

**Comment by PeterJBassett** [comment]
Hello @mgooding - I hope you can help me. I have developed a itwin connector test script based on the script you provided, and this has been working well for a couple of weeks. However today, it is giving an error 400 when it tries to do the silent login.

My test script uses the exact same code as your demo script:
```
  const authorizationClient = new NodeCliAuthorizationClient({ ...AUTH_CLIENT_CONFIG_PROPS });
  Logger.logInfo(APP_LOGGER_CATEGORY, "Attempting to sign in");
  await authorizationClient.signIn();
```
But I am getting this response today.
```
Info | iModel Test Script | Attempting to sign in 
Request ended with an error  400
```
Can you give any advice on how I can debug this?

**Reply by pmconne** [reply]
@PeterJBassett can you open the browser's dev tools and copy-paste the URL that's producing the 400 please?
EDIT: Ah never mind, I see it's a command-line app.

**Reply by PeterJBassett** [reply]
Hello @pmconne - It is not logging in from a browser, it is a silent login done using: `NodeCliAuthorizationClient `. So there is no browser dev tools in this case.

**Reply by pmconne** [reply]
Thanks, yes, I noticed that seconds after posting my silly question.
Have you tried wrapping in try-catch to get more details, e.g.,
```
try {
  await authorizationClient.signIn();
  Logger.logInfo(APP_LOGGER_CATEGORY, "Sign in successful");
} catch (e) {
  Logger.logInfo(APP_LOGGER_CATEGORY, `${e.message} ${e.stack}}`);
}

**Reply by PeterJBassett** [reply]
Thanks you I tried that and did a console.log(e) and got:
```
Request ended with an error  400
AppAuthError { message: 'Bad Request', extras: undefined }
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<AppAuthError>".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

```

**Reply by pmconne** [reply]
Ok thanks - I was hoping to see a URL or a call stack. :(
@mgooding is unavailable today, and it looks like he's the only current maintainer of the cli authorization client. Perhaps @calebmshafer has a guess what might be going wrong.

**Comment by PeterJBassett** [comment]
Hello @mgooding @calebmshafer @pmconne 
I tried registering a new app on https://developer.bentley.com/register/
I then had to copy the list of scopes into my script - this list of required scopes seems to have changed from what I had previously/
Having done this, the test script is now working again.

**Comment by PeterJBassett** [comment]
As a follow on to the last post.
In order to get the test script to work again, I had to register a new app on https://developer.bentley.com/register/ I have set the scope in the script to the minimum required scope "imodels:read".
The previous registered app does not work even with the minimum scope.

The questions are:
- Why did the previous registered app suddenly stop working?
   - Was there a change made within Bentley, or is it some sort of problem perhaps with login details being cached on my local machine?  
- How can I prevent this happening again?

**Reply by calebmshafer** [reply]
Hey @PeterJBassett, I'm not quite sure why your app suddenly stopped working. What scopes did you have registered when it stopped working? Just the minimum set of `imodels:read`?

We'll look into it on our side to see if there are any issues that stand out.

Edit: Could you also write out the exact scopes that you have registered for each of the clients on the server side?

**Reply by laurynasr** [reply]
Hi @PeterJBassett, can you please provide client_id (for both apps) and the list of scopes (also for both) you have in your app configuration for us to take a look what happened?

Thanks

**Reply by PeterJBassett** [reply]
Hi @laurynasr apologies for bthe delay replying.
The original client app I was running was https://developer.bentley.com/apps/native/native-qh93YgC5tSksGFtmasm5IT2kv
This has the following scopes that were created by the portal: "openid realitydata:read imodels:read library:read realitydata:modify library:modify storage:read storage:modify imodels:modify savedviews:modify imodelaccess:read itwinjs organization profile email savedviews:read"

The new client app I have created which works is https://developer.bentley.com/apps/native/native-CWNA0qKjz2nrTAJImRR6eto0M/
This has the scopes that were created by the portal: "library:modify library:read imodels:read realitydata:modify savedviews:modify imodels:modify storage:read realitydata:read savedviews:read storage:modify"

The only scope set in my application code is: "imodels:read".

**Reply by PeterJBassett** [reply]
Hello @calebmshafer and @laurynasr - I hope you can help. I have the same problem again that my test script application has stopped working as it is not accepting the silent login again.

The way I fixed this last time was to register a new client app, but I don't want to keep having to do this whenever the app stops working.

```
// The iTwin App configuration registered on https://developer.bentley.com/my-apps/
const AUTH_CLIENT_CONFIG_PROPS = {
  clientId: "", // EDIT ME! Specify your own clientId
  /** These are the minimum scopes needed - you can leave alone or replace with your own entries */
  scope: "imodels:read",
  /** This can be left as-is assuming you've followed the instructions in README.md when registering your application */
  redirectUri: "http://localhost:3000/signin-callback",
};
```

The silent login code is
```
const openIModelFromIModelHub = async (): Promise<BriefcaseDb> => {
  if (!AUTH_CLIENT_CONFIG_PROPS.clientId || !AUTH_CLIENT_CONFIG_PROPS.scope || !AUTH_CLIENT_CONFIG_PROPS.redirectUri)
    return Promise.reject("You must edit AUTH_CLIENT_CONFIG");

  const authorizationClient = new NodeCliAuthorizationClient({ ...AUTH_CLIENT_CONFIG_PROPS });
  Logger.logInfo(APP_LOGGER_CATEGORY, "Attempting to sign in");
  try {
  await authorizationClient.signIn();
  } catch (e: any) {
    console.log(e)
    throw(e)
  }
  Logger.logInfo(APP_LOGGER_CATEGORY, "Sign in successful");
  IModelHost.authorizationClient = authorizationClient;
  . . .
}
```

The error logged is:
```
Info | iModel Test Script | Attempting to sign in 
Request ended with an error  400
AppAuthError { message: 'Bad Request', extras: undefined }
```

**Reply by aruniverse** [reply]
This might be a bug, for the time being can you try clearing the cached token on your machine, which should hopefully request a new valid token. If youre on widnows, try opening "windows credential manager", looking for an entry with your client id, and clearing that entry. if you're on mac, it'll be in Mac OS KeyChain. Could you let us know if that works?

**Reply by PeterJBassett** [reply]
Hello @calebmshafer and @laurynasr 
Follow-up to above. I have run npm update and rebuilt, and it has started working again.
